### PR TITLE
[BUG]: Implement Kahan summation for rolling().mean() to avoid numerical issues

### LINKED ISSUE #10319: rolling_mean returns small negative value on Series with non-negative values
We came upon this issue today while trying to take a square root of a non-negative series. 

A quick search came up with this old issue: https://github.com/pydata/pandas/issues/2527.The issue was reproduced on 0.15 by tweaking the code from the previous report, which is below:

```
from pandas import date_range, Series
s = Series(index=date_range('1999-02-03','1999-04-05'))
s['1999-02-03'] = 0.00012456
s['1999-02-04'] = 0.0003
s['1999-04-04'] = -0.0
s['1999-04-05'] = -0.0
In [21]: s.rolling(1).mean()['1999-04-05']
Out[21]: -5.421010862427522e-20
```

which returns -5.4210108624275222e-20.

### LINKED ISSUE #13254: PERF/COMPAT: use kahan summation in .rolling(..).sum()
from [mailing list](https://groups.google.com/forum/#!topic/pydata/Bl7QLr-Y5Z0)

to avoid imprecision errors as the rolling computations are evaluated marginally (sliding the window and adding new / subtracting old). We tend to accumulate floating point errors.

I looks like just an extra sum and subtract so I think the perf impact would be minimal.

This might be able to be applied to `.rolling(..).mean()` as well. Note that `.rolling(..).std/var` already use Welford's algo for better precision.

https://en.wikipedia.org/wiki/Kahan_summation_algorithm

implementation from `numpy`

```
cdef double kahan_sum(double *darr, npy_intp n):
    cdef double c, y, t, sum
    cdef npy_intp i
    sum = darr[0]
    c = 0.0
    for i from 1 <= i < n:
        y = darr[i] - c
        t = sum + y
        c = (t-sum) - y
        sum = t
    return sum

```

### LINKED ISSUE #32761: Rolling mean on a time-based period with one data point does not return the exact mean
First, thank you all for creating and maintaining this package, and for the obvious care that has gone into the numerical analysis of the `rolling` methods.

#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
print('Pandas Version', pd.__version__)
pd.options.display.float_format = '{:,.15f}'.format
np.random.seed(42)

index = pd.date_range(start='2020-01-01', end= '2020-01-02', freq='1s').append(pd.DatetimeIndex(['2020-01-03']))
data = np.random.rand(len(index))

df = pd.DataFrame({'data':data}, index=index)
df['mean'] = df.rolling('60s').mean()
df
```
Result

```
>>> print('Pandas Version', pd.__version__)
Pandas Version 1.0.2
...
>>> df
                                 data              mean
2020-01-01 00:00:00 0.374540118847362 0.374540118847362
2020-01-01 00:00:01 0.950714306409916 0.662627212628639
2020-01-01 00:00:02 0.731993941811405 0.685749455689561
2020-01-01 00:00:03 0.598658484197037 0.663976712816430
2020-01-01 00:00:04 0.156018640442437 0.562385098341631
...                               ...               ...
2020-01-01 23:59:57 0.888246435024149 0.461337636515083
2020-01-01 23:59:58 0.825973496654809 0.468473569401559
2020-01-01 23:59:59 0.487629482790733 0.473417509587717
2020-01-02 00:00:00 0.705341871943242 0.480022530163534
2020-01-03 00:00:00 0.105076984427931 0.105076984428225

[86402 rows x 2 columns]
```
#### Problem description

In the example above I show a rolling mean over a 1-minute window, but the final point jumps ahead by one day. Since this is the only point in the final window, I expect the mean here to be *exactly* the single value in the window. You can see however that there is some error in the low-order 4 dights displayed.

I believe this is due to an omission in the function `roll_mean_variable` which I am reading in the file `pandas/_libs/window/aggregations.pyx`. I believe that the function should zero the running sum whenever the number of observations goes to 0 after a "remove" operation. Otherwise, some noise from previous periods will be left in the running sum.

If I apply this patch to my local copy then I get the exact mean. 

```
diff --git a/pandas/_libs/window/aggregations.pyx b/pandas/_libs/window/aggregations.pyx
index 495b436..76ba1dd 100644
--- a/pandas/_libs/window/aggregations.pyx
+++ b/pandas/_libs/window/aggregations.pyx
@@ -344,20 +344,24 @@ def roll_mean_variable(ndarray[float64_t] values, ndarray[int64_t] start,
                     val = values[j]
                     add_mean(val, &nobs, &sum_x, &neg_ct)
 
             else:
 
                 # calculate deletes
                 for j in range(start[i - 1], s):
                     val = values[j]
                     remove_mean(val, &nobs, &sum_x, &neg_ct)
 
+                # Reset sum if NOBS goes to 0
+                if nobs == 0:
+                    sum_x = 0
+
                 # calculate adds
                 for j in range(end[i - 1], e):
                     val = values[j]
                     add_mean(val, &nobs, &sum_x, &neg_ct)
 
             output[i] = calc_mean(minp, nobs, neg_ct, sum_x)
 
             if not is_monotonic_bounds:
                 for j in range(s, e):
                     val = values[j]
```

I understand that there is a philosophical question: Given that we know that most rolling means are likely not bit-accurate anyway, is it worth doing a check for this special case? One argument would be that for time series like mine with gaps here and there, the gaps would serve to clear the accumulated errors every time a gap occurred, and in general the accuracy of the procedure would improve.

I notice that the rolling variance calculations *do* reset the accumulators to 0 when they completely depopulate a window. I am curious though, is there a numerical-analysis reason for doing the adds before the removes there? This order nullifies the resetting-effect of clearing the accumulators I believe, because my windows will never go to empty with this ordering. This is the code I'm referring to: 

https://github.com/pandas-dev/pandas/blob/5da500af76d0031021a8e906d8f34dc3e6ded234/pandas/_libs/window/aggregations.pyx#L520

Thank you,

Bishop Brock

#### Expected Output

I expect the rolling mean of any window specified with a time period with a single data point to be equal to that data point. 

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.2+0.g7485dbe6f.dirty
numpy            : 1.17.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 41.4.0
Cython           : 0.29.13
pytest           : 5.2.1
hypothesis       : None
sphinx           : 2.2.0
blosc            : None
feather          : None
xlsxwriter       : 1.2.1
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.8.0
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.0
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.2.1
pyxlsb           : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.9
tables           : 3.5.2
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.1
numba            : 0.45.1

</details>

### LINKED ISSUE #36031: BUG: Mean of zeros turn out to be non-zero
- [ ] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---
#### Code Sample, a copy-pastable example

```python
'''Two examples of a time-weighted rolling average, both of which should result in a final value of 0.'''
import pandas as pd


def succeeding_test():
    df = pd.DataFrame(
        {"A": [300239975158033.0, -0.0, -0.0]},
        index=[
            pd.Timestamp("19700101 09:00:00"),
            pd.Timestamp("19700101 09:00:03"),
            pd.Timestamp("19700101 09:00:06"),
        ],
    )
    v = df.resample("1s").ffill().rolling("3s", closed="left", min_periods=3).mean().values[-1]
    assert v == 0


def failing_test():
    df = pd.DataFrame(
        {"A": [3002399751580331.0, -0.0, -0.0]},  # First value is a single digit longer.
        index=[
            pd.Timestamp("19700101 09:00:00"),
            pd.Timestamp("19700101 09:00:03"),
            pd.Timestamp("19700101 09:00:06"),
        ],
    )
    v = df.resample("1s").ffill().rolling("3s", closed="left", min_periods=3).mean().values[-1]
    assert v == 0  # v turns out to be == -0.3333333
```

#### Problem description

In both these cases, the final rolling mean value is the mean of three zeros, but in the `failing_test`, this does not result in a zero value. I am expecting the `failing_test` to have the same behaviour as the `succeeding_test`. The difference lies in the magnitude of the first value in the df, but that shouldn't affect the last mean, as it falls outside the rolling window for that value.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2ca0a2665b2d169c97de87b8e778dbed86aea07
python           : 3.6.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-42-generic
Version          : #46-Ubuntu SMP Fri Jul 10 00:24:02 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.1
numpy            : 1.16.4
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.2
setuptools       : 46.1.1
Cython           : None
pytest           : 4.6.11
hypothesis       : 4.57.1
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.8.4 (dt dec pq3 ext lo64)
jinja2           : 2.11.2
IPython          : 7.16.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 1.0.1
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.13
tables           : 3.6.1
tabulate         : None
xarray           : 0.16.0
xlrd             : 1.2.0
xlwt             : None
numba            : 0.47.0

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #36031
- [x] closes #32761
- [x] closes #10319 
- [x] xref #11645 (fixes the issue expect the 0, will add an additional pr about the doc)
- [x] closes #13254
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I implemented the Kahan summation as suggested by @jreback. I used the variable names from https://en.wikipedia.org/wiki/Kahan_summation_algorithm. If there exists a name convention am not aware of, I am happy to rename the variables.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
i believe this is shared code with mean; so pls test that as well

**Comment by phofl** [comment]
Which mean function are you referring to?

rolling.mean() is the test for and groupby.mean() and DataFrame.mean() are not running through the code.

**Comment by phofl** [comment]
Done

**Comment by mroeschke** [comment]
I dont think this hits `resample.mean`. I think @jreback might have meant just `rolling.mean`.

Could you do the same for `rolling.sum`? https://github.com/pandas-dev/pandas/issues/13254

**Comment by phofl** [comment]
@mroeschke 
Yeah you are right. I don't know what I have seen that made me believe ``resample.mean()`` would use the same function. Thanks.

``rolling.sum()`` uses another Cython function. That was not fixed with my implementation. I would be happy to implement Kahan summation there too. Should I open another PR or mix it in here?

**Comment by jreback** [comment]
> i believe this is shared code with mean; so pls test that as well

sorry I meant rolling.sum

yeah ideally we could share code, though these are simple enough and we don't want to introduce functions that it may be just easier to duplicate what is needed.

ok on this PR or followon to do rolling.sum

File: pandas/_libs/window/aggregations.pyx [hunk 1]
```diff
@@ -277,32 +277,42 @@ cdef inline float64_t calc_mean(int64_t minp, Py_ssize_t nobs,
 
 
 cdef inline void add_mean(float64_t val, Py_ssize_t *nobs, float64_t *sum_x,
-                          Py_ssize_t *neg_ct) nogil:
-    """ add a value from the mean calc """
+                          Py_ssize_t *neg_ct, float64_t *c) nogil:
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx:280]: can you make c a more meaningful label

File: pandas/_libs/window/aggregations.pyx [hunk 2]
```diff
@@ -277,32 +277,42 @@ cdef inline float64_t calc_mean(int64_t minp, Py_ssize_t nobs,
...
+    """ add a value from the mean calc using Kahan summation """
+    cdef:
+        float64_t y, t
 
     # Not NaN
     if notnan(val):
         nobs[0] = nobs[0] + 1
-        sum_x[0] = sum_x[0] + val
+        y = val - c[0]
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx:288]: add a comment that using kahan summation

File: pandas/tests/window/test_rolling.py [hunk 3]
```diff
@@ -696,3 +696,21 @@ def scaled_sum(*args):
     expected = DataFrame(data={"X": [0.0, 0.5, 1.0, 1.5, 2.0]}, index=_index)
     result = df.groupby(**grouping).rolling(1).apply(scaled_sum, raw=raw, args=(2,))
     tm.assert_frame_equal(result, expected)
+
+
+def test_rolling_numerical_accuracy_kahan():
+    # GH: 36031 implementing kahan summation
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py:702]: can you add the other test from the OP (the succeeding one) as well, pls parameterize.

**Review by jreback — CHANGES_REQUESTED** [review]
for this PR or more likely kahan sum, I think there are a couple of issues on the tracker. if you'd have a look can link them all (and use examples from them).

File: pandas/_libs/window/aggregations.pyx (hunk 2 above)
Comment (phofl) [reply pandas/_libs/window/aggregations.pyx:288]: I already added a comment in the function header. Is this sufficient?

File: pandas/_libs/window/aggregations.pyx (hunk 1 above)
Comment (phofl) [reply pandas/_libs/window/aggregations.pyx]: Done

File: pandas/tests/window/test_rolling.py (hunk 3 above)
Comment (phofl) [reply pandas/tests/window/test_rolling.py:703]: Done

**Comment by phofl** [comment]
@jreback Implemented Kahan summation for rolling().sum(). I also added tests for the issues I've found.

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -105,6 +105,7 @@ Other enhancements
 - :class:`Index` with object dtype supports division and multiplication (:issue:`34160`)
 - :meth:`DataFrame.explode` and :meth:`Series.explode` now support exploding of sets (:issue:`35614`)
 - `Styler` now allows direct CSS class name addition to individual data cells (:issue:`36159`)
+- :meth:`Series.rolling().mean()`, :meth:`DataFrame.rolling().mean()` :meth:`Series.rolling().sum()` and :meth:`DataFrame.rolling().sum()`use Kahan summation to calculate the mean to avoid numerical problems (:issue:`10319`, :issue:`11645`, :issue:`13254`, :issue:`32761`, :issue:`36031`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.2.0.rst:108]: Use ``` :meth:`Rolling.mean` ``` and ``` :meth:`Rolling.sum` ``` instead

File: pandas/tests/window/test_rolling.py [hunk 5]
```diff
@@ -696,3 +696,62 @@ def scaled_sum(*args):
...
+def test_rolling_numerical_accuracy_jump():
+    # GH: 32761
+    index = pd.date_range(start="2020-01-01", end="2020-01-02", freq="1s").append(
+        pd.DatetimeIndex(["2020-01-03"])
+    )
+    data = np.random.rand(len(index))
+
+    df = pd.DataFrame({"data": data}, index=index)
+    df["mean"] = df.rolling("60s").mean()
+    assert df["mean"][-1] == df["data"][-1]
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:735]: For this test let's test the whole dataframe with `tm.assert_frame_equal`

File: pandas/tests/window/test_rolling.py [hunk 6]
```diff
@@ -696,3 +696,62 @@ def scaled_sum(*args):
...
+
+
+def test_rolling_numerical_accuracy_small_values():
+    # GH: 10319
+    s = Series(
+        data=[0.00012456, 0.0003, -0.0, -0.0],
+        index=date_range("1999-02-03", "1999-02-06"),
+    )
+    assert s.rolling(1).mean()["1999-02-06"] == 0.0
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:744]: Same here except with `tm.assert_series_equal`

**Comment by phofl** [comment]
Will look into Test failures

File: pandas/tests/window/test_rolling.py (hunk 6 above)
Comment (phofl) [reply pandas/tests/window/test_rolling.py]: Done

File: pandas/tests/window/test_rolling.py (hunk 5 above)
Comment (phofl) [reply pandas/tests/window/test_rolling.py]: done

File: doc/source/whatsnew/v1.2.0.rst (hunk 4 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst]: Thanks, that is way better.

**Comment by phofl** [comment]
Tests should be fixed now. Somehow mixed plus and minus up

File: pandas/tests/window/test_rolling.py [hunk 7]
```diff
@@ -696,3 +696,63 @@ def scaled_sum(*args):
...
+        index=[
+            pd.Timestamp("19700101 09:00:00"),
+            pd.Timestamp("19700101 09:00:03"),
+            pd.Timestamp("19700101 09:00:06"),
+        ],
+    )
+    result = (
+        df.resample("1s").ffill().rolling("3s", closed="left", min_periods=3).mean()
+    )
+    assert result.values[-1] == 0.0
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:715]: Could we compare the entire DataFrame result for this test as well?

**Review by mroeschke — COMMENTED** [review]
One more comment otherwise LGTM

**Review by jreback — CHANGES_REQUESTED** [review]
ok looks good, ex @mroeschke comments.

can you run all of the rolling asv's and report the results. I would expect a small slowdown.

File: pandas/tests/window/test_rolling.py (hunk 7 above)
Comment (phofl) [reply pandas/tests/window/test_rolling.py]: Done

**Comment by phofl** [comment]
@jreback 

I ran asv for rolling. Is the output within the range you expected?

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl very nice!

yeah this is a sligth perf hit, but worth the tradeoff

**Pull request merged**